evm/p2p/cancel_token.py

```diff
@@ -0,0 +1,90 @@
+import asyncio
+from typing import Any, Awaitable
+
+from evm.p2p.exceptions import OperationCancelled
+
+
+class CancelToken:
+
+    def __init__(self, name: str, next_token: 'CancelToken' = None) -> None:
+        self.name = name
+        self.next = next_token
+        self._triggered = asyncio.Event()
+
+    def chain(self, next_token: 'CancelToken') -> 'CancelToken':
+        """Return a new CancelToken chaining this and next_token.
+
+        The new CancelToken's triggered will return True if trigger() has been
+        called on either of the chained tokens, but calling trigger() on this token
+        has no effect on either of the chained tokens.
+        """
+        return CancelToken(":".join([self.name, next_token.name]), next_token)
```

Comment:
It's confusing to me that the resulting token seems to have no reference to `self` (besides the name of course). Shouldn't chaining be just something like `self.next = next_token`?

Reply:
Good catch; this is indeed a bug! It was not caught by the tests and didn't affect our code because they were always triggering the last token on the chain. I've added more tests and changed `CancelToken` to keep a list of chained tokens instead of a single reference to the next item on the chain, as when chaining we want to create a new token that can be triggered separately from the original one